src/Conversion/ONNXToKrnl/NN/Conv.cpp

```diff
@@ -16,12 +15,228 @@
+                        createKrnl.getLoc(), tmpType);
+                createKrnl.store(fZero, reductionVal);
+
+                // Bounds for reduction loops.
+                ValueRange redLoops = createKrnl.defineLoops(spacialRank + 1);
+                SmallVector<IndexExpr, 4> redLbs, redUbs;
+                // First: loop over channel in per group.
+                redLbs.emplace_back(iZero);
+                redUbs.emplace_back(SymbolIndexExpr(CIPerGroup));
+                // For each spacial dim, do the following.
+                for (int i = 0; i < spacialRank; ++i) {
+                  // Get data for dis spacial dimension.
+                  DimIndexExpr o(outputSpatialIndices[i]);
+                  SymbolIndexExpr I(
+                      inputBounds.getSymbol(spatialStartIndex + i));
+                  SymbolIndexExpr K(
+                      filterBounds.getSymbol(spatialStartIndex + i));
+                  LiteralIndexExpr p(pads[i]); // Begining/left/top pad.
+                  LiteralIndexExpr s(strides[i]);
+                  LiteralIndexExpr d(isDilated ? dilations[i] : 1);
+                  // lb = ceil((p - o * s) / d)
+                  IndexExpr pos = p - (o * s);
+                  IndexExpr lb = pos.ceilDiv(d);
+                  lb = IndexExpr::max(lb, 0);
+                  redLbs.emplace_back(lb);
+                  // ub = ceil((I + p - o * s) / d)
+                  IndexExpr ipos = I + pos;
+                  IndexExpr ub = ipos.ceilDiv(d);
+                  ub = IndexExpr::min(ub, K);
+                  redUbs.emplace_back(ub);
+                }
+                // for ciPerGroup = 0 .. CIPerGroup:
+                //   for kh in lb .. ub:
+                //     for kw in lb .. ub:
+                createKrnl.iterateIE(redLoops, redLoops, redLbs, redUbs, {},
+                    [&](KrnlBuilder &createKrnl, ValueRange) {
+                      ValueRange redIndices =
+                          createKrnl.getInductionVarValue(redLoops);
+                      IndexExprScope redScope(createKrnl);
+                      MathBuilder createMath(createKrnl);
+                      // Create access function for input image:
+                      // [n, ci, ho * sh + kh * dh - ph, wo * sw + kw * dw -
+                      // pw].
+                      SmallVector<IndexExpr, 4> inputAccessFct;
+                      DimIndexExpr n(outerIndices[0]);
+                      inputAccessFct.emplace_back(n);
+                      // ci = g * CIPerG + ciPerG
+                      DimIndexExpr ciPerG(redIndices[0]);
+                      IndexExpr ci =
+                          DimIndexExpr(g) * SymbolIndexExpr(CIPerGroup) +
+                          ciPerG;
+                      inputAccessFct.emplace_back(ci);
+                      for (int i = 0; i < spacialRank; ++i) {
+                        // for each spacial dims: access is o * s + k * d - p.
```

Comment:
Is it better to move the computation for `o * s - p` outside the current loop? Perhaps, a similar movement for `ci = g * CIPerG + ciPerG`

Reply:
Updated the code as requested. Not sure it is not caught by later optimizations in the backend, but it probably does not hurt to do it here.

Reply:
Thank you for the update!